### PR TITLE
Counters

### PR DESCRIPTION
**Pull request by azazana**
Check deviation for newCounter

### CODE REVIEW

File: generator/counter.go [hunk 1]
```diff
@@ -11,7 +14,8 @@ type Counter struct {
 // NewCounter returns new generator for growing points. Starting value is an increment as well.
 // Possibly it can randomize values around increment.
 // When deviation is set, it the next value won't be less then previous.
-func NewCounter(name string, start, stop, step uint, randomizeStart bool, value, deviation float64) *Counter {
+func NewCounter(name string, start, stop, step uint, randomizeStart bool, value, deviation float64) (*Counter, error) {
+
```
Comment (Felixoid) [inline generator/counter.go:18]: Why empty lines in the beginning and ending of functions?

**Review by Felixoid — CHANGES_REQUESTED** [review]
Few minor things to go

File: generator/counter.go (hunk 1 above)
Comment (azazana) [reply generator/counter.go:18]: I have fixed it

File: generator/counter.go [hunk 2]
```diff
@@ -24,8 +28,12 @@ func NewCounter(name string, start, stop, step uint, randomizeStart bool, value,
 		},
 		increment: value,
 	}
+	if deviation != 0 && value < 0 && math.Abs(deviation) > math.Abs(value) {
```
[the remark this answers is not in the document]
Comment (azazana) [reply generator/counter.go:31]: 1. I have to return a counter or can I avoid it?
2. Ок
3. Ок

[the remark this answers is not in the document]
Comment (Felixoid) [reply generator/counter.go]: By convenient, usually the empty object returned together with error. Bill in the current case

File: generator/counter.go [hunk 3]
```diff
@@ -15,6 +15,9 @@ type Counter struct {
 // Possibly it can randomize values around increment.
 // When deviation is set, it the next value won't be less then previous.
 func NewCounter(name string, start, stop, step uint, randomizeStart bool, value, deviation float64) (*Counter, error) {
+	if value < 0 && math.Abs(value) < math.Abs(deviation) {
+		return &Counter{}, ErrNewCounter
```
Comment (Felixoid) [inline generator/counter.go:19]: nil will be better, it won't create an object

Comment (Felixoid) [reply generator/counter.go:19]:
```suggestion
		return nil, ErrNewCounter
```

Comment (azazana) [reply generator/counter.go]: Ok

File: generator/counter_test.go [hunk 4]
```diff
@@ -20,16 +20,22 @@ func TestCounterNew(t *testing.T) {
...
+		c, e = NewCounter("metric.name", 12, 15, 100, true, 30, 0)
 		if c.Time() != 12 {
 			randomized = true
 			break
 		}
 	}
+	assert.NoError(t, e)
 	assert.True(t, randomized)
+
+	_, e = NewCounter("metric.name", 12, 15, 1, false, -30, 31)
```
Comment (Felixoid) [inline generator/counter_test.go:36]: And now, here we can check for `assert.Nil` :-j

File: generator/base.go [hunk 5]
```diff
@@ -59,9 +59,9 @@ func GetType(typeName string) (Type, error) {
 	return t, nil
 }
 
-func (b *base) getNextCommon() error {
+func (b *base) getTime() error {
 	if b.time > b.stop {
-		return ErrGenOver
+		return fmt.Errorf("the last point reached")
```
Comment (Felixoid) [inline generator/base.go:64]: Nope. it should be `ErrGenOver`

File: generator/const.go [hunk 6]
```diff
@@ -30,7 +30,7 @@ func NewConst(name string, start, stop, step uint, randomizeStart bool, value, d
 
 // Next sets value and time for the next point
 func (c *Const) Next() error {
-	if c.getNextCommon() != nil {
+	if c.getTime() != nil {
```
Comment (Felixoid) [inline generator/const.go:33]: The previous variant `err := c.getTime(); err != nil` was correct, but the following line should be `return err`

File: generator/base.go [hunk 7]
```diff
@@ -59,6 +59,14 @@ func GetType(typeName string) (Type, error) {
 	return t, nil
 }
 
+func (b *base) getTime() error {
```
Comment (Felixoid) [inline generator/base.go:62]:
```suggestion
func (b *base) nextTime() error {
```

File: generator/counter.go [hunk 8]
```diff
@@ -11,7 +14,10 @@ type Counter struct {
 // NewCounter returns new generator for growing points. Starting value is an increment as well.
 // Possibly it can randomize values around increment.
 // When deviation is set, it the next value won't be less then previous.
-func NewCounter(name string, start, stop, step uint, randomizeStart bool, value, deviation float64) *Counter {
+func NewCounter(name string, start, stop, step uint, randomizeStart bool, value, deviation float64) (*Counter, error) {
+	if value < 0 && math.Abs(value) <= math.Abs(deviation) {
+		return nil, ErrNewCounter
```
Comment (Felixoid) [inline generator/counter.go:19]:
```suggestion
		return nil, fmt.Errorf("%w: with negative value deviation must be greater than value", ErrNewCounter)
```

File: generator/interface.go [hunk 9]
```diff
@@ -52,7 +55,7 @@ func New(typeName, name string, start, stop, step uint, randomizeStart bool, val
 	case ConstType:
 		return NewConst(name, start, stop, step, randomizeStart, value, deviation), nil
```
Comment (Felixoid) [inline generator/interface.go:56]: May be a good option to have consistent New* functions

**Comment by Felixoid** [comment]
@azazana tests failed :-j

**Pull request merged**